include/ringmesh/basic/attributes.h

```diff
@@ -0,0 +1,1024 @@
+        {
+            return store_;
+        }
+
+        /**
+         * \brief Gets a property value
+         * \param[in] i the index of the item
+         * \return the value of the property, convertex
+         *  into a double
+         * \pre is_bound() && i < size()
+         */
+        virtual double operator[]( index_t i ) = 0;
+
+        /**
+         * \brief Gets the number of scalar components per item in an
+         *  AttributeStore.
+         * \param[in] store a pointer to the attribute store.
+         * \return the number of scalar components per item in an
+         *  AttributeStore.
+         */
+        virtual index_t nb_scalar_elements_per_item() const = 0;
+        virtual bool is_integer_like_attribute() const = 0;
+
+    protected:
+        /**
+         * \brief Gets the base attribute name from a compound name.
+         * \param[in] name the string with the attribute name and optional
+         *  index. For instance, "foobar[5]" refers to the 5th coordinate of
+         *  the "foobar" vector attribute.
+         * \return the attribute name. For instance, for "foobar[5]", it
+         *  returns "foobar".
+         */
+        static std::string attribute_base_name( const std::string& name );
+
+        /**
+         * \brief Gets the base attribute name from a compound name.
+         * \param[in] name the string with the attribute name and optional
+         *  index. For instance, "foobar[5]" refers to the 5th coordinate of
+         *  the "foobar" vector attribute.
+         * \return the index or zero if no index was specified. For instance,
+         *  for "foobar[5]" it returns 5, and for "foobar" it returns 0
+         */
+        static index_t attribute_element_index( const std::string& name );
+
+        /**
+         * \brief Gets the element type stored in an AttributeStore.
+         * \param[in] store a const pointer to the AttributeStore
+         * \return one of ET_UINT8, ET_INT8, ET_UINT32, ET_INT32,
+         *  ET_FLOAT32, ET_FLOAT64 if the type of the attribute is
+         *  compatible with those types, or ET_NONE if it is incompatible.
+         */
+        // static ElementType element_type( const AttributeStore* store );
+
+        /**
+         * \brief Gets an element.
+         * \param[in] i index of the element
+         * \param[in] multiplier multiplier applied to the index before fetching
+         *  the raw pointer.
+         */
+        template < class T >
+        double get_element( index_t i, index_t multiplier = 1 ) const
+        {
+            ringmesh_unused( multiplier );
+            ringmesh_assert( is_bound() );
+            ringmesh_assert( i < size() );
+            auto value = *reinterpret_cast< const T* >( store_->element( i ) );
+            return double( value );
+        }
+
+    protected:
+        const AttributesManager* manager_;
+        const AttributeStore* store_;
+        index_t element_index_;

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Make this member variable "private". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3656)